streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -294,11 +306,13 @@ public synchronized void start() {
         log.debug("Starting Kafka Stream process");
 
         if (state == KafkaStreams.State.CREATED) {
-            for (final StreamThread thread : threads)
-                thread.start();
-
             setState(KafkaStreams.State.RUNNING);
 
+            for (final StreamThread thread : threads) {
```

Comment:
A thread can immediately go into a state (e.g., `ASSIGNING_PARTITIONS`) and call the state listener which then sets the `KafkaStreams.State` to `REBALANCING`. If I revert the order, we would be overwriting that state with `RUNNING`.